src/cocotb/share/lib/verilator/verilator.cpp

```diff
@@ -113,32 +114,27 @@ int main(int argc, char** argv) {
 #endif
 
     while (!Verilated::gotFinish()) {
-        // Call registered timed callbacks (e.g. clock timer)
-        // These are called at the beginning of the time step
-        // before the iterative regions (IEEE 1800-2012 4.4.1)
-        VerilatedVpi::callTimedCbs();
-
-        // Call Value Change callbacks triggered by Timer callbacks
-        // These can modify signal values
-        settle_value_callbacks();
-
-        // We must evaluate whole design until we process all 'events'
-        bool again = true;
-        while (again) {
-            // Evaluate design
+        // We must evaluate whole design until we process all 'events' for this
+        // time step
+        do {
             top->eval_step();
+            VerilatedVpi::clearEvalNeeded();
 
-            // Call Value Change callbacks triggered by eval()
-            // These can modify signal values
-            again = settle_value_callbacks();
+            // Call Value Change callbacks
+            // These can modify signal values so we loop
+            // until there are no more changes
+            bool again;
+            do {
+                VerilatedVpi::doInertialPuts();
+                again = VerilatedVpi::callValueCbs();
```

Comment:
Hmmmm... Swapping these actually fixes both test errors. What I don't understand is why...

Reply:
I see... 

1. A timer callback (the clock) wakes up before eval. Writes are queued. 
2. Eval occurs for the beginning of time step / any timing processes.
3. Writes are flushed (the clock change)
4. Value change callbacks (`RisingEdge(clk)`) are run. Writes are queued.
5. If any value change callbacks were run (they were), the writes are applied and the loop runs again (!!! without an eval !!!).

So the issue is the clock writes are applied, `RisingEdge(clk)` returns which queues writes, those writes are applied immediately.

What we need is:
```python
while True:
    doTimedCbs()
    eval()
    while True:
        doInertialPuts()
        doValueChangeCbs()
        if not evalNeeded():
            break
        eval()
    doReadWriteCbs()
```

Reply:
And technically the first eval is only needed if we are in `--timing` mode, because otherwise there is no time-based events in the HDL to wake up.

Reply:
Turns out there was not actually a problem. The inertial_writes_tests failures were due to me getting mixed up, and the test_dff failure was just due to how Verilator works after the change.